dm/relay/binlog_writer.go

```diff
@@ -105,32 +176,46 @@ func (w *BinlogWriter) Close() error {
 
 	w.file = nil
 	w.offset.Store(0)
-	w.uuid = ""
-	w.filename = ""
+	w.uuid.Store("")
+	w.filename.Store("")
+	w.input = nil
 
+	if writeErr := w.Error(); writeErr != nil {
+		return writeErr
+	}
 	return err
 }
 
 func (w *BinlogWriter) Write(rawData []byte) error {
-	w.mu.RLock()
-	defer w.mu.RUnlock()
-
 	if w.file == nil {
 		return terror.ErrRelayWriterNotOpened.Delegate(errors.New("file not opened"))
 	}
+	w.input <- rawData
+	w.offset.Add(int64(len(rawData)))
+	return w.Error()
```

Comment:
since every Write need to check Error, Error is a hot function and we'd better using fast implementation. I prefer atomic variable like using `atomic.Error` and CAS to replace nil to an error